trzmq/_socket.py

```diff
@@ -349,37 +39,26 @@ def _update_events(self, wake_all=False):
         if wake_all or self._events & zmq.POLLOUT:
             self._wake(zmq.POLLOUT)
 
-    @contextlib.contextmanager
-    def _cancel_when(self, flag):
-        with trio.open_cancel_scope() as cancel_scope:
-            try:
-                self._cancel_scopes[flag].add(cancel_scope)
-                yield
-            finally:
-                self._cancel_scopes[flag].remove(cancel_scope)
-
-    @trio.hazmat.enable_ki_protection
+    @trio.lowlevel.enable_ki_protection
     async def _wait_for(self, flag):
         if self._events & flag:
-            await trio.hazmat.yield_briefly()
+            await trio.lowlevel.checkpoint()
             return
 
         while not self._events & flag:
-            with self._cancel_when(flag):
-                if not self._someone_is_watching_trigger:
-                    self._someone_is_watching_trigger = True
-                    try:
-                        await trio.hazmat.wait_socket_readable(self._trigger)
-                    finally:
-                        self._someone_is_watching_trigger = False
-                        # Wake up everyone, regardless of whether their event
-                        # is ready, b/c they need to potentially take over the
-                        # job of sitting in wait_socket_readable.
-                        self._update_events(wake_all=True)
-                else:
-                    # Someone else is watching the socket, so we just need to
-                    # wait to be woken up.
-                    await trio.sleep_forever()
+            if not self._someone_is_watching_trigger:
+                self._someone_is_watching_trigger = True
+                try:
+                    await trio.lowlevel.wait_readable(self._zmq_sock.fd)
+                finally:
+                    self._someone_is_watching_trigger = False
+                    self._update_events()
```

Comment:
The old code had `wake_all=True` here and a comment explaining why -- is that still needed or are you handling that somewhere else somehow?

Reply:
I had convinced myself it wasn't needed, because the rule is that anyone who wakes up will do whatever they wanted to do and then call _update_events.

I was just writing an example to show that was right, and instead I think I've come up with an example that DOES need wake_all, so I guess I'm wrong.

If all the waiters are wanting to write: one will end up watching the trigger and one will end up sleeping. If the watcher wakes up, it will do its write, and then call _update_events which doesn't wake up the other waiter, and now no one is listening for the trigger.

In short, I think you're right and I should add it back.

Reply:
No, that's not correct.

In that example the other waiter would be awoken (the event is broadcast to all of the same type of waiters) and end up as the next watcher when it discovers the socket isn't writable anymore.

So that means the case of "watcher and waiter are the same type" is ok.

For the case of "watcher and waiter are of different types". The watcher will wake up: if the event matches the waiter, it's obviously ok because the watcher will go back to watching. If the event matches the watcher, the watcher will leave _wait_for and temporarily no one is watching the trigger. The watcher then does whatever it wanted to do and then calls _update_events.

My assertion is that this will always cause the waiter to be awoken (if the watcher wrote, then the waiter must have been waiting for a message, and it will now see that there is one to read. If the watcher read, then the waiter must have been waiting for space to write a message and it will now be available.)

Reply:
Is that taking into account that the reason the watcher might wake up is b/c they were cancelled?